astropy/convolution/convolve.py

```diff
@@ -410,13 +441,20 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0, crop=True,
 
     # mask catching - masks must be turned into NaNs for use later
     if np.ma.is_masked(array):
-        mask = array.mask
+        mamask = array.mask
+        array = np.array(array)
+        array[mamask] = np.nan
+    elif mask is not None:
+        # copying here because we have to mask it below.  But no need to copy
+        # if mask is None because we won't modify it.
         array = np.array(array)
-        array[mask] = np.nan
+    if mask is not None:
+        # mask != 0 yields a bool mask for all ints/floats/bool
+        array[mask != 0] = np.nan
     if np.ma.is_masked(kernel):
-        mask = kernel.mask
+        kmask = kernel.mask
         kernel = np.array(kernel)
-        kernel[mask] = np.nan
+        kernel[kmask] = np.nan
```

Comment:
It occurred to me this is exactly what `MaskedArray.filled(val)` is for, so it's now an identical one-liner in both functions!